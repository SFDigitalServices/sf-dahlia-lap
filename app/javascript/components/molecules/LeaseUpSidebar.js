import React, { useState } from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'

import LeaseUpSidebarButtons from './LeaseUpSidebarButtons'
import StatusItems from './StatusItems'

import StatusItemShape from '../../utils/shapes/StatusItemShape'

const MAX_UPDATES_TO_SHOW_DEFAULT = 4

const getMostRecentStatus = (statusHistory) =>
  statusHistory && statusHistory[0]
    ? statusHistory[0].status
    : null

// visible for testing
export const ShowOrHideStatusesButton = ({ showingAllStatuses, onClick }) => {
  const buttonClasses = classNames(
    'button-link',
    't-tiny',
    'show-all-updates-toggle'
  )

  return (
    <button
      className={buttonClasses}
      type='button'
      onClick={onClick}
    >
      {showingAllStatuses ? 'Show only recent status updates' : 'Show all status updates'}
    </button>
  )
}

const LeaseUpSidebar = ({ isLoading, statusItems, onSaveClicked, onChangeStatus, onAddCommentClicked }) => {
  const [showingAllStatuses, setShowingAllStatuses] = useState(false)

  const currentStatus = getMostRecentStatus(statusItems)

  const sidebarButtons = (withMobileStyling) => (
    <div className={withMobileStyling ? 'hide-large-up' : 'show-large-up'}>
      <LeaseUpSidebarButtons
        isLoading={isLoading}
        status={currentStatus}
        withMobileStyling={withMobileStyling}
        onSaveClicked={onSaveClicked}
        onChangeStatus={onChangeStatus}
        onAddCommentClicked={onAddCommentClicked}
      />
    </div>
  )
  const onShowHideStatusesToggled = () => setShowingAllStatuses(!showingAllStatuses)

  const numberOfStatusesToDisplay = showingAllStatuses ? statusItems.length : MAX_UPDATES_TO_SHOW_DEFAULT

  return (
    <div className='sidebar-content'>
      {sidebarButtons(false)}
      <h3 className='status-history-label'>Status History</h3>
      <StatusItems
        statusItems={statusItems}
        limit={numberOfStatusesToDisplay}
      />
      {statusItems.length > MAX_UPDATES_TO_SHOW_DEFAULT && (
        <ShowOrHideStatusesButton
          showingAllStatuses={showingAllStatuses}
          onClick={onShowHideStatusesToggled}
        />
      )}
      {sidebarButtons(true)}
    </div>
  )
}

LeaseUpSidebar.propTypes = {
  isLoading: PropTypes.bool,
  statusItems: PropTypes.arrayOf(PropTypes.shape(StatusItemShape)),
  onSaveClicked: PropTypes.func,
  onChangeStatus: PropTypes.func,
  onAddCommentClicked: PropTypes.func
}

LeaseUpSidebar.defaultProps = {
  isLoading: false,
  statusItems: [],
  onSaveClicked: null,
  onChangeStatus: null,
  onAddCommentClicked: null
}

export default LeaseUpSidebar
