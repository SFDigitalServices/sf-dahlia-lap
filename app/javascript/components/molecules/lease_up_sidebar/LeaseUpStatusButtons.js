import React from 'react'

import PropTypes from 'prop-types'

import { LEASE_UP_STATUS_VALUES } from '../../../utils/statusUtils'
import StatusDropdown from '../StatusDropdown'

const LeaseUpStatusButtons = ({ status, isLoading, onChangeStatus, onAddCommentClicked }) => (
  <div className='status-history-buttons'>
    <StatusDropdown
      status={status}
      onChange={onChangeStatus}
      disabled={isLoading}
      buttonClasses={['tight-padding']}
    />
    <button
      // ID is just for e2e tests
      id='add-status-history-comment'
      className='tertiary tight-padding'
      disabled={isLoading}
      type='button'
      onClick={onAddCommentClicked}
    >
      Add a comment
    </button>
  </div>
)

LeaseUpStatusButtons.propTypes = {
  status: PropTypes.oneOf(LEASE_UP_STATUS_VALUES),
  isLoading: PropTypes.bool,
  onChangeStatus: PropTypes.func,
  onAddCommentClicked: PropTypes.func
}

LeaseUpStatusButtons.defaultProps = {
  status: null,
  isLoading: false,
  onChangeStatus: null,
  onAddCommentClicked: null
}

export default LeaseUpStatusButtons
