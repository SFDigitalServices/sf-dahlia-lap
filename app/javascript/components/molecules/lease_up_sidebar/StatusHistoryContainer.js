import React, { useState } from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'

import StatusItems from './StatusItems'

import StatusItemShape from '../../../utils/shapes/StatusItemShape'
import ContentSection from '../ContentSection'

const MAX_UPDATES_TO_SHOW_DEFAULT = 4

/**
 * Displays a list of status items in a container with a header and
 * a show/hide entire status item list link.
 *
 * Defaults to only showing 4 status items at a time.
 */
const StatusHistoryContainer = ({ statusItems }) => {
  const [showingAllStatuses, setShowingAllStatuses] = useState(false)

  const onShowHideStatusesToggled = () =>
    setShowingAllStatuses(!showingAllStatuses)

  const numberOfStatusesToDisplay = showingAllStatuses
    ? statusItems.length
    : MAX_UPDATES_TO_SHOW_DEFAULT

  return (
    <>
      <div className='padding-top--3x show-large-up'>
        <ContentSection.SubHeader title='Status History' />
      </div>
      <div className='padding-top--half hide-large-up'>
        <ContentSection.Header title='Status History' />
      </div>
      <StatusItems
        statusItems={statusItems}
        limit={numberOfStatusesToDisplay}
      />
      {statusItems.length > MAX_UPDATES_TO_SHOW_DEFAULT && (
        <button
          className={classNames('button-link', 't-tiny', 'show-all-updates-toggle')}
          type='button'
          onClick={onShowHideStatusesToggled}
        >
          {showingAllStatuses
            ? 'Show only recent status updates'
            : 'Show all status updates'}
        </button>
      )}
    </>
  )
}

StatusHistoryContainer.propTypes = {
  statusItems: PropTypes.arrayOf(PropTypes.shape(StatusItemShape)).isRequired
}

export default StatusHistoryContainer
