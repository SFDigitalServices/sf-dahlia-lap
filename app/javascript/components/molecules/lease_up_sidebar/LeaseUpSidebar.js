import React from 'react'

import { isEmpty } from 'lodash'
import PropTypes from 'prop-types'

import StatusItemShape from '../../../utils/shapes/StatusItemShape'
import LeaseUpStatusButtons from './LeaseUpStatusButtons'
import StatusHistoryContainer from './StatusHistoryContainer'

const getMostRecentStatus = (statusHistory) =>
  statusHistory && statusHistory[0] ? statusHistory[0].status : null

const LeaseUpSidebar = ({
  isLoading,
  statusItems,
  onSaveClicked,
  onChangeStatus,
  onAddCommentClicked
}) => {
  const currentStatus = getMostRecentStatus(statusItems)

  const saveButton = (
    <button
      className='primary save-button'
      id='save-supplemental-application'
      type='submit'
      disabled={isLoading}
      onClick={onSaveClicked}
    >
      {isLoading ? 'Savingâ€¦' : 'Save'}
    </button>
  )

  const statusButtons = (
    <LeaseUpStatusButtons
      isLoading={isLoading}
      status={currentStatus}
      onChangeStatus={onChangeStatus}
      onAddCommentClicked={onAddCommentClicked}
    />
  )

  return (
    <div className='sidebar-content'>
      <div className={'show-large-up'}>
        {saveButton}
        {statusButtons}
      </div>
      {!isEmpty(statusItems) && <StatusHistoryContainer statusItems={statusItems} />}
      <div className={'hide-large-up'}>
        {statusButtons}
        {saveButton}
      </div>
    </div>
  )
}

LeaseUpSidebar.propTypes = {
  isLoading: PropTypes.bool,
  statusItems: PropTypes.arrayOf(PropTypes.shape(StatusItemShape)),
  onSaveClicked: PropTypes.func,
  onChangeStatus: PropTypes.func,
  onAddCommentClicked: PropTypes.func
}

LeaseUpSidebar.defaultProps = {
  isLoading: false,
  statusItems: [],
  onSaveClicked: null,
  onChangeStatus: null,
  onAddCommentClicked: null
}

export default LeaseUpSidebar
