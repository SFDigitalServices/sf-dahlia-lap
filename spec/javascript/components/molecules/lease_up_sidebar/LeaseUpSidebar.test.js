import React from 'react'
import { shallow } from 'enzyme'
import LeaseUpSidebar from 'components/molecules/lease_up_sidebar/LeaseUpSidebar'
import LeaseUpStatusButtons from 'components/molecules/lease_up_sidebar/LeaseUpStatusButtons'
import StatusHistoryContainer from 'components/molecules/lease_up_sidebar/StatusHistoryContainer'
import { mockStatusItem, mockStatusItems, mockManyStatusItems } from '../../../mocks/statusItemMock'

const getWrapper = (items, isLoading = false) => shallow(
  <LeaseUpSidebar
    statusItems={items}
    isLoading={isLoading}
  />
)

describe('LeaseUpSidebar', () => {
  describe('when not loading', () => {
    test('should render with empty status items correctly', () => {
      const wrapper = getWrapper([])
      // two sidebar buttons components, one for mobile, one for desktop
      expect(wrapper.find(LeaseUpStatusButtons)).toHaveLength(2)
      expect(wrapper.find(LeaseUpStatusButtons).first().prop('status')).toBeNull()
      expect(wrapper.find(StatusHistoryContainer)).toHaveLength(0)
    })

    test('should render with a single status item correctly', () => {
      const wrapper = getWrapper([mockStatusItem()])
      expect(wrapper.find(LeaseUpStatusButtons)).toHaveLength(2)
      expect(wrapper.find(LeaseUpStatusButtons).first().prop('status')).toEqual('Approved')
      expect(wrapper.find(LeaseUpStatusButtons).first().prop('isLoading')).toBeFalsy()
      expect(wrapper.find(StatusHistoryContainer).prop('statusItems')).toHaveLength(1)
    })

    test('should render with multiple status items correctly', () => {
      const wrapper = getWrapper(mockStatusItems())
      expect(wrapper.find(LeaseUpStatusButtons)).toHaveLength(2)
      expect(wrapper.find(LeaseUpStatusButtons).first().prop('status')).toEqual('Approved')
      expect(wrapper.find(StatusHistoryContainer).prop('statusItems')).toHaveLength(6)
    })

    test('should render with more than 4 status items correctly', () => {
      const wrapper = getWrapper(mockManyStatusItems(5))
      expect(wrapper.find(LeaseUpStatusButtons)).toHaveLength(2)
      expect(wrapper.find(LeaseUpStatusButtons).first().prop('status')).toEqual('Approved')
      expect(wrapper.find(StatusHistoryContainer).prop('statusItems')).toHaveLength(5)
    })

    test('should render with lease signed status', () => {
      const wrapper = getWrapper([
        mockStatusItem({
          status: 'Lease Signed',
          substatus: null
        }),
        mockStatusItem()
      ])

      expect(wrapper.find(LeaseUpStatusButtons)).toHaveLength(2)
      expect(wrapper.find(LeaseUpStatusButtons).first().prop('status')).toEqual('Lease Signed')
    })
  })

  describe('when loading', () => {
    test('should set the buttons to be loading/disabled', () => {
      const wrapper = getWrapper([mockStatusItem()], true)

      const statusButtonsWrapper = wrapper.find(LeaseUpStatusButtons).first()
      const saveButtonWrapper = wrapper.find('button#save-supplemental-application').first()

      expect(statusButtonsWrapper.prop('isLoading')).toBeTruthy()
      expect(saveButtonWrapper.prop('disabled')).toBeTruthy()
    })
  })

  describe('snapshot tests', () => {
    test('should match snapshot with 5 status items', () => {
      const wrapper = getWrapper(mockManyStatusItems(5))
      expect(wrapper).toMatchSnapshot()
    })
  })
})
