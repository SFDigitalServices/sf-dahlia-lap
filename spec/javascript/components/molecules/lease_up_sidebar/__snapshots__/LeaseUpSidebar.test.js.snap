// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LeaseUpSidebar snapshot tests should match snapshot with 5 status items 1`] = `
<div
  className="sidebar-content"
>
  <div
    className="show-large-up"
  >
    <button
      className="primary save-button"
      disabled={false}
      id="save-supplemental-application"
      onClick={null}
      type="submit"
    >
      Save
    </button>
    <LeaseUpStatusButtons
      isLoading={false}
      onAddCommentClicked={null}
      onChangeStatus={null}
      status="Approved"
    />
  </div>
  <StatusHistoryContainer
    statusItems={
      Array [
        Object {
          "comment": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque bibendum condimentum lorem consectetur eleifend.",
          "status": "Approved",
          "substatus": "Approval letter sent",
          "timestamp": 1598385600,
        },
        Object {
          "comment": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque bibendum condimentum lorem consectetur eleifend.",
          "status": "Approved",
          "substatus": "Approval letter sent",
          "timestamp": 1598385600,
        },
        Object {
          "comment": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque bibendum condimentum lorem consectetur eleifend.",
          "status": "Approved",
          "substatus": "Approval letter sent",
          "timestamp": 1598385600,
        },
        Object {
          "comment": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque bibendum condimentum lorem consectetur eleifend.",
          "status": "Approved",
          "substatus": "Approval letter sent",
          "timestamp": 1598385600,
        },
        Object {
          "comment": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque bibendum condimentum lorem consectetur eleifend.",
          "status": "Approved",
          "substatus": "Approval letter sent",
          "timestamp": 1598385600,
        },
      ]
    }
  />
  <div
    className="hide-large-up"
  >
    <LeaseUpStatusButtons
      isLoading={false}
      onAddCommentClicked={null}
      onChangeStatus={null}
      status="Approved"
    />
    <button
      className="primary save-button"
      disabled={false}
      id="save-supplemental-application"
      onClick={null}
      type="submit"
    >
      Save
    </button>
  </div>
</div>
`;
